acceptorwrapper/rate_limiting_conn.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) TFG Co. All Rights Reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package acceptorwrapper
+
+import (
+	"container/list"
+	"net"
+	"time"
+
+	"github.com/topfreegames/pitaya"
+	"github.com/topfreegames/pitaya/constants"
+	"github.com/topfreegames/pitaya/logger"
+	"github.com/topfreegames/pitaya/metrics"
+)
+
+// RateLimitingConn wraps net.Conn by applying rate limiting
+// and return empty if exceeded
+type RateLimitingConn struct {
+	net.Conn
+	limit        int
+	interval     time.Duration
+	times        list.List
+	forceDisable bool
+}
+
+// NewRateLimitingConn returns an initialized *RateLimiting
+func NewRateLimitingConn(
+	conn net.Conn,
+	limit int,
+	interval time.Duration,
+	forceDisable bool,
+) *RateLimitingConn {
+	r := &RateLimitingConn{
+		Conn:         conn,
+		limit:        limit,
+		interval:     interval,
+		forceDisable: forceDisable,
+	}
+
+	r.times.Init()
+
+	return r
+}
+
+func (r *RateLimitingConn) Read(b []byte) (n int, err error) {
+	if r.forceDisable {
+		return r.Conn.Read(b)
+	}
+
+	n, err = r.Conn.Read(b)
+	if err != nil {
+		return n, err
+	}
+
+	now := time.Now()
+	err = r.take(now)
+	if err != nil {
+		logger.Log.Errorf("Data=%s, Error=%s", b, err)
+		metrics.ReportExceededRateLimiting(pitaya.GetMetricsReporters())
+		return 0, nil
```

Comment:
I'm not sure you want to return here, the default behaviour of net.Conn is that it only returns when it has data to return, returning here you're changing this behaviour for the ones who are consuming a wrapped connection, isn't it better to have this inside a while and have it continue on errors?

Reply:
Ok, that makes sense

Reply:
But on the other hand, from the client's point of view the messages are simply getting lost, the client isn't notified about the rate limiting happening, this might lead to hard to detect bugs.

Reply:
I'm not sure if the client should be notified, in what situations would a client hit the limiting?
1) he's cheating/trying to attack us
2) we inserted a but in clients code
in both case I think that it's irrelevant for us to answer anything for them, what do you guys think?

Reply:
@codingllama: any thoughts?

Reply:
I don't think it would be hard to detect because it's reported using metricsReporters and logged as an error

Reply:
I'm not sure I have enough context here. For example:

* Is dropping this data OK could it cause bugs?
* Does rate-limiting at the connection level implies a 1:1 relationship between a Read and RPC, or could this cause corrupted reads?
* Could we rate limit at a higher abstraction level? (eg, per RPC)

Reply:
- The effect of dropping would be a timeout error on the client side. So it depends on how the game deals with it. Maybe expliciting this behaviour on docs? So they can be prepared for this case. 
- Currently, Read can have more than one RPC but never less (we don't fragment packets now). Even though we can have more than one, we think it's a good approximation to rate limit based on Read
- Yes we could. But in my discussion with @felipejfc we thought it was better to not mix the logic afterwards and rate limit as soon as possible. Meaning that we don't need to decode the packets to rate limit it.